### PR TITLE
Add contributions utilities module and use it in 3 tests

### PR DESCRIPTION
**Pull request by desbo**
## What does this change?
Adds a module that abstracts most of the common code in contributions A/B tests. I've created 2 new tests that make use of this module, and updated the `always-ask` test to use it as well.

The module also adds some functionality to limit how many times we show the Epic (6 times a week) and wires up the targeting tool to the tests, so we can control the tags without raising other PRs.

I've also included a small `geolocation` module that saves some more duplicate code and caches the result to save on network requests.

## What is the value of this and can you measure success?
There's a lot less code in contributions A/B tests, meaning future tests will be easier to write, and their PRs significantly easier to review. In this case a test was about 40% of its former size (180 down to 70 lines).

## Does this affect other platforms - Amp, Apps, etc?
Nope

### CODE REVIEW

File: static/src/javascripts/projects/common/modules/commercial/contributions-utilities.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        viewLog.push({
+            date: new Date().getTime(),
+            testId: testId
+        });
+
+        storage.local.set(viewKey, viewLog);
+    }
+
+    function canShow() {
+        var window = maxViews.days * 1000 * 60 * 60 * 24;
```
Comment (gtrufitt) [inline static/src/javascripts/projects/common/modules/commercial/contributions-utilities.js]: Maybe window is a bit of a dodgy `var` name?

File: static/src/javascripts/projects/common/modules/experiments/tests/contributions-epic-always-ask-strategy.js [hunk 2]
```diff
@@ -108,13 +110,29 @@ define([
         };
 
         var canBeDisplayed = function() {
+
+            var isCharityAskPage = function() {
+                var charityKeywordIds = [
+                    // TODO
```
Comment (desbo) [inline static/src/javascripts/projects/common/modules/experiments/tests/contributions-epic-always-ask-strategy.js]: I've just realised I need to fix this. Won't be long

Comment (gtrufitt) [inline static/src/javascripts/projects/common/modules/experiments/tests/contributions-epic-always-ask-strategy.js]: This ok?

File: static/src/javascripts/projects/common/modules/experiments/tests/contributions-epic-always-ask-strategy.js [hunk 3]
```diff
@@ -108,13 +110,29 @@ define([
...
+                ];
+                var pageKeywordIdsString = config.page.keywordIds;
+                if (typeof (pageKeywordIdsString) !== 'undefined') {
+                    var pageKeywordIds = pageKeywordIdsString.split(',');
+                    return intersection(charityKeywordIds, pageKeywordIds).length > 0;
+                } else {
+                    return false;
```
Comment (gtrufitt) [inline static/src/javascripts/projects/common/modules/experiments/tests/contributions-epic-always-ask-strategy.js]: No need for the else branch (this'll be flagged in new linting)

**Review by gtrufitt — APPROVED** [review]
A couple of superficial questions, but LGTM 👍

File: static/src/javascripts/projects/common/modules/commercial/contributions-utilities.js [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    function daysSince(date) {
+        var oneDay = 24 * 60 * 60 * 1000;
+
+        try {
+            var ms = Date.parse(date);
+
+            if (isNaN(ms)) return Infinity;
+            else return (new Date() - ms) / oneDay;
```
Comment (gtrufitt) [inline static/src/javascripts/projects/common/modules/commercial/contributions-utilities.js]: No need for the `else`

**Comment by desbo** [comment]
thank you gareth! 🙌  
will address your comments before merging

**Comment by jranks123** [comment]
👍 👊

File: static/src/javascripts/projects/common/utils/geolocation.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    fetch
+) {
+    var location;
+
+    return {
+        get: function() {
+            return new Promise(function(resolve, reject) {
+                if (location) return resolve(location);
+                else {
+                    fetch('https://api.nextgen.guardianapps.co.uk/geolocation', {
```
Comment (alexduf) [inline static/src/javascripts/projects/common/utils/geolocation.js]: that would be nice to use `config.page.ajaxUrl` so when we test locally it goes to localhost

**Comment by alexduf** [comment]
That's going to make things really easy for contributions, thanks @desbo !

**Pull request merged**